cpp/velox/utils/GpuBufferBatchResizer.cc

```diff
@@ -0,0 +1,346 @@
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+        chars.data(), valueBuffer->data_as<uint8_t>(), chars.size(), cudaMemcpyDefault, stream.value()));
+    return cudf::make_strings_column(
+        numRows, std::move(offsetColumn), std::move(chars), nullCount, std::move(*mask.release()));
+  }
+};
+
+template <>
+std::unique_ptr<cudf::column> DispatchColumn::readFlatColumn<TypeKind::VARCHAR>(cudf::type_id typeId) {
+  return readFlatColumnStringView(typeId);
+}
+
+template <>
+std::unique_ptr<cudf::column> DispatchColumn::readFlatColumn<TypeKind::VARBINARY>(cudf::type_id typeId) {
+  return readFlatColumnStringView(typeId);
+}
+
+std::shared_ptr<VeloxColumnarBatch> makeCudfTable(
+    RowTypePtr type,
+    int32_t numRows,
+    const std::vector<std::shared_ptr<arrow::Buffer>>& buffers,
+    memory::MemoryPool* pool) {
+  std::vector<std::unique_ptr<cudf::column>> cudfColumns;
+  cudfColumns.reserve(type->size());
+
+  int32_t bufferIdx = 0;
+  auto stream = cudf_velox::cudfGlobalStreamPool().get_stream();
+  DispatchColumn dispatch{stream, cudf::get_current_device_resource_ref(), std::move(buffers), bufferIdx, numRows};
+  for (const auto& colType : type->children()) {
+    auto res = VELOX_DYNAMIC_SCALAR_TYPE_DISPATCH(
+        dispatch.readFlatColumn, colType->kind(), cudf_velox::veloxToCudfTypeId(colType));
+    cudfColumns.emplace_back(std::move(res));
+  }
+  auto cudfTable = std::make_unique<cudf::table>(std::move(cudfColumns));
+  stream.synchronize();
+  return std::make_shared<VeloxColumnarBatch>(
+      std::make_shared<cudf_velox::CudfVector>(pool, type, numRows, std::move(cudfTable), stream));
+}
+
+} // namespace
+
+GpuBufferBatchResizer::GpuBufferBatchResizer(
+    arrow::MemoryPool* arrowPool,
+    facebook::velox::memory::MemoryPool* pool,
+    int32_t minOutputBatchSize,
+    std::unique_ptr<ColumnarBatchIterator> in)
+    : arrowPool_(arrowPool),
+      pool_(pool),
+      minOutputBatchSize_(minOutputBatchSize),
+      in_(std::move(in)) {
+  VELOX_CHECK_GT(minOutputBatchSize_, 0, "minOutputBatchSize should be larger than 0");
+}
+
+std::shared_ptr<ColumnarBatch> GpuBufferBatchResizer::next() {
+  std::vector<std::shared_ptr<GpuBufferColumnarBatch>> cachedBatches;
+  int32_t cachedRows = 0;
+  while (cachedRows < minOutputBatchSize_) {
+    auto nextCb = in_->next();
+    if (!nextCb) {
+      // No more input.
+      break;
+    }
+
+    auto nextBatch = std::dynamic_pointer_cast<GpuBufferColumnarBatch>(nextCb);
+    VELOX_CHECK_NOT_NULL(nextBatch);
+    if (nextBatch->numRows() == 0) {
+        continue;
+    }
+
+    cachedRows += nextBatch->numRows();
+    cachedBatches.push_back(std::move(nextBatch));
+  }
+  if (cachedRows == 0) {
+    return nullptr;
+  }
+
+  // Compose all cached batches into one
+  auto batch = GpuBufferColumnarBatch::compose(arrowPool_, cachedBatches, cachedRows);
+
+  auto finalBatch = makeColumnarBatch(batch->getRowType(), cachedRows, batch->buffers(), pool_);
```

Comment:
Hi, @devavret @karthikeyann I'm trying to deserialize to cudf::table instead of velox RowVector, when I use L335 makeColumnarBatch to construct a velox RowVector, Q95 succeeds,  but if I change to L338, the result is incorrect(I update CudfFromVelox local to also accept CudfVector as input https://github.com/oap-project/velox/compare/2025_10_20...jinchengchenghh:velox:2025_10_20), could you help review to find what is wrong? All the implementation is in this file. Much Thanks! I also write a small unit test to check if the makeCudfTable returned table is equal to RowVector, the tests can pass.